### PR TITLE
clarify account lockout message

### PR DESCRIPTION
**Pull request by caseylitton**
Needed changes for testing:
1. lms/envs/devstack.py 
-FEATURES['ENABLE_MAX_FAILED_LOGIN_ATTEMPTS'] = True

Optional changes for testing:
1. lms/envs/private.py
-MAX_FAILED_LOGIN_ATTEMPTS_ALLOWED = 5 #lower for for testing, I used 5.
-MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS = 15 * 60 #lower for for testing, I used 15 * 1

### CODE REVIEW

**Comment by caseylitton** [comment]
@stvstnfrd @kluo please review.  I still nee to clean up commit message and add testing instructions.

File: common/djangoapps/student/models.py [hunk 1]
```diff
@@ -794,7 +794,7 @@ def is_user_locked_out(cls, user):
             now = datetime.now(UTC)
             until = record.lockout_until
             is_locked_out = until and now < until
-
+            print 'is_locked_out', until and now < until #cl
```
Comment (stvstnfrd) [inline common/djangoapps/student/models.py]: This is a debug statement; we should remove it.

File: common/djangoapps/student/models.py [hunk 2]
```diff
@@ -807,9 +807,10 @@ def increment_lockout_counter(cls, user):
         record, _ = LoginFailures.objects.get_or_create(user=user)
         record.failure_count = record.failure_count + 1
         max_failures_allowed = settings.MAX_FAILED_LOGIN_ATTEMPTS_ALLOWED
-
+        print 'record.failure_count', record.failure_count  #cl
```
Comment (stvstnfrd) [inline common/djangoapps/student/models.py]: This is a debug statement; we should remove it.

File: common/djangoapps/student/models.py [hunk 3]
```diff
@@ -807,9 +807,10 @@ def increment_lockout_counter(cls, user):
...
         # did we go over the limit in attempts
         if record.failure_count >= max_failures_allowed:
+            print 'locked out' #cl
```
Comment (stvstnfrd) [inline common/djangoapps/student/models.py]: This is a debug statement; we should remove it.

File: lms/envs/common.py [hunk 4]
```diff
@@ -2310,7 +2310,6 @@
 MAX_FAILED_LOGIN_ATTEMPTS_ALLOWED = 5
 MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS = 15 * 60
 
-
```
Comment (stvstnfrd) [inline lms/envs/common.py]: We should avoid making unrelated changes; back this out.

File: lms/envs/devstack.py [hunk 5]
```diff
@@ -129,8 +129,12 @@ def should_show_debug_toolbar(_):
 
 ########################## SECURITY #######################
 FEATURES['ENFORCE_PASSWORD_POLICY'] = False
-FEATURES['ENABLE_MAX_FAILED_LOGIN_ATTEMPTS'] = False
-FEATURES['SQUELCH_PII_IN_LOGS'] = False
+
+#FEATURES['ENABLE_MAX_FAILED_LOGIN_ATTEMPTS'] = False
+FEATURES['ENABLE_MAX_FAILED_LOGIN_ATTEMPTS'] = True #cl debug improve_account_lockout_messaging
```
Comment (stvstnfrd) [inline lms/envs/devstack.py]: You could place this in your `lms/envs/private.py` file to avoid temporarily checking it in. Either way, you'll want to remove this before merging.

File: common/djangoapps/student/views.py [hunk 6]
```diff
@@ -1284,9 +1284,12 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
     user_found_by_email_lookup = user
     if user_found_by_email_lookup and LoginFailures.is_feature_enabled():
         if LoginFailures.is_user_locked_out(user_found_by_email_lookup):
+            overest_lockout_mins=str(int(settings.MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS*.0222+1)) #cl .0222 is 33% more with the conversion into minutes.  1.33/60=0.02216
             return JsonResponse({
                 "success": False,
-                "value": _('This account has been temporarily locked due to excessive login failures. Try again later.'),
+                "value": _('This account has been temporarily locked due to excessive login failures. '
+                            'Try again in '+overest_lockout_mins+' minutes.  For security reasons, '
+                            'reseting the password will NOT lift the lockout. Please wait for '+overest_lockout_mins+' minutes.'),
```
Comment (stvstnfrd) [inline common/djangoapps/student/views.py]: You'll want to use string formating instead of manual concatenation.

Also, be sure to translate the message first, and _then_ substitute parameters.

As discussed F2F, this should look something like:

```python
"value": _(
    "This account has been temporarily locked due to excessive login failures. "
    "Try again in {minutes} minutes.  For security reasons, "
    "reseting the password will NOT lift the lockout. Please wait for {minutes} minutes."
).format(
   minutes=overest_lockout_mins,
),
```

File: common/djangoapps/student/views.py [hunk 7]
```diff
@@ -1284,9 +1284,12 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
     user_found_by_email_lookup = user
     if user_found_by_email_lookup and LoginFailures.is_feature_enabled():
         if LoginFailures.is_user_locked_out(user_found_by_email_lookup):
+            overest_lockout_mins=str(int(settings.MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS*.0222+1)) #cl .0222 is 33% more with the conversion into minutes.  1.33/60=0.02216
```
Comment (stvstnfrd) [inline common/djangoapps/student/views.py]: Since we're calling `format` now, we can drop the cast to `str`.

Comment (stvstnfrd) [inline common/djangoapps/student/views.py]: Since this value never changes, we should avoid recalculating the math.
Let's move this up to the top level as a local constant.

```python
OVERESTIMATE_LOCKOUT_MINUTES = int(settings.MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS*.0222+1)
```

**Comment by caseylitton** [comment]
@stvstnfrd @kluo went through Steven's initial comments, ready for round 2

File: common/djangoapps/student/views.py [hunk 8]
```diff
@@ -1284,9 +1285,16 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
     user_found_by_email_lookup = user
     if user_found_by_email_lookup and LoginFailures.is_feature_enabled():
         if LoginFailures.is_user_locked_out(user_found_by_email_lookup):
+            overest_lockout_mins=str(int(settings.MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS*.0222+1)) #cl .0222 is 33% more with the conversion into minutes.  1.33/60=0.022166
```
Comment (caesar2164) [inline common/djangoapps/student/views.py]: remove

File: lms/envs/devstack.py [hunk 9]
```diff
@@ -129,8 +129,9 @@ def should_show_debug_toolbar(_):
 
 ########################## SECURITY #######################
 FEATURES['ENFORCE_PASSWORD_POLICY'] = False
-FEATURES['ENABLE_MAX_FAILED_LOGIN_ATTEMPTS'] = False
+FEATURES['ENABLE_MAX_FAILED_LOGIN_ATTEMPTS'] = True
 FEATURES['SQUELCH_PII_IN_LOGS'] = False
+
```
Comment (caesar2164) [inline lms/envs/devstack.py]: remove added newline

File: common/djangoapps/student/models.py [hunk 10]
```diff
@@ -795,7 +795,7 @@ def is_user_locked_out(cls, user):
             until = record.lockout_until
             is_locked_out = until and now < until
 
-            return is_locked_out
```
Comment (caesar2164) [inline common/djangoapps/student/models.py]: ?

File: common/djangoapps/student/models.py [hunk 11]
```diff
@@ -810,6 +810,7 @@ def increment_lockout_counter(cls, user):
 
         # did we go over the limit in attempts
         if record.failure_count >= max_failures_allowed:
+
```
Comment (caesar2164) [inline common/djangoapps/student/models.py]: remove

**Comment by caesar2164** [comment]
awesome, the code changes look great!

now you need to modify the breaking test so it passes!

File: common/djangoapps/student/views.py [hunk 12]
```diff
@@ -141,6 +141,7 @@
 ReverifyInfo = namedtuple('ReverifyInfo', 'course_id course_name course_number date status display')  # pylint: disable=invalid-name
 SETTING_CHANGE_INITIATED = 'edx.user.settings.change_initiated'
 
+MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_OVER_ESTIMATE_MINUTES=int(settings.MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS*.0222+1) #0.0222 is 33% more with the conversion into minutes.  1.33/60=0.02216
```
Comment (stvstnfrd) [inline common/djangoapps/student/views.py]: This approach was really clever and I appreciate that it was your initial instinct. The more I think about it, this really should be a case where simple trumps clever. While the 30% calculation provides the expected outcome given this input (15min), it starts to fall apart if we ever change the timeout.

If we drop the timeout to 10m, the message will say 14.
If we raise the timeout to 20m, the message will say 27.

We wouldn't want to display an error message with odd numbers like that; we want something divisible by 5.
```python
DISPLAY_LOGIN_LOCKOUT_MINUTES = 5 + settings.MAX_FAILED_LOGIN_ATTEMPS_LOCKOUT_PERIOD_SECS
```

If we bump the timeout to 60m, we're adding 20 minutes to the wait time, instead of just 5.

Really, the intent here is to grant us just a little bit of leeway, in case they click even a few seconds early. 

5 minutes all the time should be plenty.

File: cms/djangoapps/contentstore/tests/tests.py [hunk 13]
```diff
@@ -23,6 +23,7 @@
 
 from freezegun import freeze_time
 
+from student.view import MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_OVER_ESTIMATE_MINUTES
```
Comment (caesar2164) [inline cms/djangoapps/contentstore/tests/tests.py]: this should probably be `student.views`

File: common/djangoapps/student/views.py [hunk 14]
```diff
@@ -141,7 +141,7 @@
 ReverifyInfo = namedtuple('ReverifyInfo', 'course_id course_name course_number date status display')  # pylint: disable=invalid-name
 SETTING_CHANGE_INITIATED = 'edx.user.settings.change_initiated'
 
-MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_PLUS_FIVE_MINUTES = 5 + settings.MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS
+LOGIN_LOCKOUT_PERIOD_PLUS_FIVE_MINUTES = int((5 * 60 + settings.MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS) / 60)
```
Comment (caesar2164) [inline common/djangoapps/student/views.py]: `5 + int(settings.MAX_FAILED_LOGIN_ATTEMPTS_LOCKOUT_PERIOD_SECS)/60` seems equivalent

**Comment by caseylitton** [comment]
@stvstnfrd @kluo final review

**Comment by stvstnfrd** [comment]
+1

**Comment by caesar2164** [comment]
👍  (even though you didn't simplify the formula 😉 )

**Comment by kluo** [comment]
+1
